Change program name in the localisation strings

DimitarSamarov07 opened the pull request:
<!--

IMPORTANT! READ

Any spam PRs will be closed.
Please make sure your PR
complies with the requirements.
-->

Resolves: *#145 (partially)*

- Update the locale files from Weblate
- Update all relevant strings containing "Audacity" to "Tenacity" in audacity.pot and the rest of the files in locale/
- Introduce a patch which solves configuration issues for Mac, regarding the locales(credit to @imachug for providing the patch)


<!-- Use "x" to fill the checkboxes below like [x]
an asterisk (*) after the entry indicates required
-->

- [x] I have signed off my commits using `-s` or `Signed-off-by`\* (See: [Contributing § DCO](https://github.com/tenacityteam/tenacity/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code\*
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving\*
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"\*

\* indicates required

fossdd commented:
@emabrey Is the weblate project not there for such purposes?

DimitarSamarov07 commented:
@fossdd  I want to note that I am aware of its existence, but it's way easier this way + it's a 2 way sync between weblate and the git repo so I see no problem with me just making a pull request.

fossdd commented:
Yes, you're right. 'replace-all' is a nice thing which doesn't exist in weblate.

DimitarSamarov07 commented:
Currently considering updating Audacity in all locales. It will prob take long if I have to do it all manually so I am thinking of using some kind of Regex combined with Replace All to do it for me and after that I can review the changes. I suck at Regex but maybe sth like this can do the job:

`/(?<=msgstr ".*\b)(Audacity)(?=\b.*")/gi`

After that I can maybe design sth that is gonna replace it in non-path related comments? Either way, working on it.

emabrey commented on locale/audacity.pot:
```diff
@@ -1,13 +1,13 @@
-"Project-Id-Version: audacity 3.0.3\n"
-"Report-Msgid-Bugs-To: audacity-translation@lists.sourceforge.net\n"
+"Project-Id-Version: tenacity 3.0.3\n"
+"Report-Msgid-Bugs-To: tenacity-translation@lists.sourceforge.net\n"
```
This isn't our bug report location for translations. Perhaps https://github.com/tenacityteam/tenacity/issues instead?

SFR-git replied:
seeing as it's an email address, I propose `~tenacity/tenacity-discuss@lists.sr.ht`

DimitarSamarov07 replied:
Good point. Gonna fix it when I get home.

DimitarSamarov07 replied:
> seeing as it's an email address, I propose `~tenacity/tenacity-discuss@lists.sr.ht`

I think that until the project gets an actual email address with the sole purpose of reporting translation issues, putting a link to issues would be better

Semisol replied:
I would preferably put it as tenacity-discuss.

DimitarSamarov07 replied:
@Semisol @emabrey Went with the majority, changed the email to _tenacity-discuss@lists.sr.ht_

emabrey commented:
We can't use weblate to change the english strings because they are used directly. We need to split it into an `en.po` file before we can use weblate for this kind of change. We can change every other language file on Weblate, but not weblate. Also, audacity.pot is a GENERATED file, so as soon as a build gets done I'm like 99% sure this is all going to be obliterated anyways. We can use `msgmerge` to reverse generate PO files from this POT file, but that is fraught with pitfalls. This change should wait until we split out an `en.po` file.

DimitarSamarov07 commented:
> We can't use weblate to change the english strings because they are used directly. We need to split it into an `en.po` file before we can use weblate for this kind of change. We can change every other language file on Weblate, but not English. Also, `audacity.pot` is a GENERATED file, so as soon as a build gets done I'm like 99% sure this is all going to be obliterated anyways. We can use `msgmerge` to reverse generate PO files from this POT file, but that is fraught with pitfalls. This change should wait until we split out an `en.po` file.

At this point, will it help if I replace all the occurrences of "Audacity" in the translation files for the other languages?

Semisol reviewed (dismissed):
Note that it is <code><b><i>~tenacity</i></b>/tenacity-discuss@lists.sr.ht</code>

emabrey commented:
> 
> 
> Yes, you're right. 'replace-all' is a nice thing which doesn't exist in weblate.

You can use tools/search and replace  I believe. Worked for me during testing.

DimitarSamarov07 commented:
> We can't use weblate to change the english strings because they are used directly. We need to split it into an `en.po` file before we can use weblate for this kind of change. We can change every other language file on Weblate, but not English. Also, `audacity.pot` is a GENERATED file, so as soon as a build gets done I'm like 99% sure this is all going to be obliterated anyways. We can use `msgmerge` to reverse generate PO files from this POT file, but that is fraught with pitfalls. This change should wait until we split out an `en.po` file. 

I think that updating the `audacity.pot` will be in everyone's best interest as the source strings in weblate will be updated which will make it less confusing for the translators. If you really take a look, a lot of the .po files have been generated a pretty long time ago which to me means they aren't regenerated in every new build. Notice how it says that the .po file was generated by audacity.pot in some files, but it doesn't say it in others. Running the script for updating files always adds this line. I am nearly finished with replacing Audacity with Tenacity in all locale translations so I will prob rename the pull request when I finish with that.

Semisol reviewed (commented):
LGTM, I am not sure if this is 100% correct though so someone else should review it

falkTX commented:
Wasn't the intention to use Englsh translation file to do this?

Won't this change basically break all translation strings that use "audacity"?

DimitarSamarov07 commented:
> Wasn't the intention to use Englsh translation file to do this?
> 
> Won't this change basically break all translation strings that use "audacity"?

It won't because I will also correct them. Gonna commit when I get home later. This is NOT a permanent solution, but it will make it easier for the translators until there's a more permanent solution(e.g an En file or actually renaming stuff in the code). I'm also going to put the translations that we currently have on weblate so we should be good to go(I hope)

falkTX commented:
do we have stuff to translate already? it seems to be time is better spent getting the English translation file to work.

what happens after we have an english translation? we revert this PR?

DimitarSamarov07 commented:
@Semisol what do you think of the idea of me creating a shell script(in a separate PR) to update the locales in the repo from Weblate and maybe you or someone from the maintainers can configure it to work with Github Actions and for e.g update the locales every 24hrs or so?

DimitarSamarov07 commented:
> do we have stuff to translate already? it seems to be time is better spent getting the English translation file to work.
> 
> what happens after we have an english translation? we revert this PR?

Or maybe we copy paste the stuff from the English part of this PR into a new file because the audacity.pot is just a normal translation file? Not sure what's your point.

falkTX commented:
Point is to reduce changes from upstream audacity.
This PR changes a core file for translation, which I imagine is going to get changed by upstream a lot.

Thinking more about this, maybe it is for the best. if we have a en.po not available upstream, we won't catch any code changes to it and thus be forced to fix it.

DimitarSamarov07 commented:
@Semisol @emabrey Finally, changed the program name in all the translation strings(long live RegExp). Renamed the PR. All the files are up-to date with Weblate's as I replaced them manually. Let me know if you want me to change anything.

panos commented:
There seems to be a series of merge conflicts here? @DimitarSamarov07

DimitarSamarov07 commented:
> There seems to be a series of merge conflicts here? @DimitarSamarov07 

Seems like a commit in master yesterday caused this to happen because some files were removed while others were renamed or straight up removed. Either way, working on it

DimitarSamarov07 commented:
@panos Resolving this conflict was a huge pain. Either way, I created a new branch, fetched tenacity's master branch, resolved the conflicts and merged to my master branch, Should be good to go now. Let me know what you think.

panos commented:
> Should be good to go now. Let me know what you think.

Hi, excuse me for the late response.

Strings such as "About Audacity..." in the Help menu and some references to Audacity in the About menu are still there. Are you sure this went alright? Also, why is the Mac build failing?

panos commented:
Also, we **cannot** get rid of the credit meant for Audacity's development team from our source files.

DimitarSamarov07 commented:
> > Should be good to go now. Let me know what you think.
> 
> Hi, excuse me for the late response.
> 
> Strings such as "About Audacity..." in the Help menu and some references to Audacity in the About menu are still there. Are you sure this went alright? Also, why is the Mac build failing?

Mac build is not failing because of the translations, you can check the build log. It started when I manually merged master into my branch to resolve the conflicts, so that's sth which is not caused by me. If you look through issues and PRs, you will see a lot complaining about issues with Mac builds. I changed **nothing** which could cause the build to fail. I've only made changes to locales. Regarding the mentioned strings, I may have missed some. Could you please refer to the file where you see this problem so I can fix it ? It's possible that the fix won't work for the English localisation because audacity.pot is auto-generated(not sure if this is done manually or runtime) but I edited it either way(but I think it should work). Important thing is that the locales for the other languages work which I think is a step in the right direction. It will make the Translators' job easier.

panos commented:
Got it, yeah, you seem to be right. I'll look into this deeper. I'll admit that I'm not completely sure what's happening because the Audacity name is so deeply embedded to the point where things just randomly break due to unforeseeable reasons. Could you rebase your commits? That could *probably* work.

(For future reference, instead of merging the master branch, you can always run `git pull origin master --rebase`.)

panos commented:
If the macOS tests fail, that most likely means that people using macOS won't be able to compile the program and that's a problem. You could temporarily remove the change that breaks things, but we will still have to deal with it later down the line. Restricting the scope of this change is probably a good idea, however. We can always fix remaining references later.

DimitarSamarov07 commented:
> If the macOS tests fail, that most likely means that people using macOS won't be able to compile the program and that's a problem. You could temporarily remove the change that breaks things, but we will still have to deal with it later down the line. Restricting the scope of this change is probably a good idea, however. We can always fix remaining references later.

Got ya. It was my fault for not merging properly, rarely had to deal with such a large amount of conflicts so I messed up. Gonna revert the non-locale files when I get home later, sorry for the inconvenience.

panos commented:
> Got ya. It was my fault for not merging properly, rarely had to deal with such a large amount of conflicts so I messed up. Gonna revert the non-locale files when I get home later, sorry for the inconvenience.

No problem, it's more about the fact that things haven't been stabilized yet. Feel free to give it another shot.

DimitarSamarov07 commented:
@panos Think I nailed it this time. Noticed a few more stuff that I have to change in the locale files but either way the build shouldn't fail now.

DimitarSamarov07 commented:
@panos I think I addressed all the mentioned above problems. Not sure if the Mac build is gonna past this time, but the Arch linux one seems to be passing, so that means the program itself **should** work. The problem which was causing the build to fail was a non-existent localization file mentioned in LINGUAS.  Let me know if you want me to address sth else.

DimitarSamarov07 commented:
Alright, checked everything countless times. I see no reason to why the mac build would fail. I haven't changed anything that can impact this. Would be grateful if someone can help me out to see what's wrong but I don't think it's anything that I did. Gonna also test stuff with the original files locally, but I doubt it's a problem with the locales as the build only fails on Mac which leads me to believe it's sth mac-specific

imachug commented:
Try to apply this patch

```patch
diff --git a/locale/CMakeLists.txt b/locale/CMakeLists.txt
index 052418ff6..ab7443e4e 100755
--- a/locale/CMakeLists.txt
+++ b/locale/CMakeLists.txt
@@ -61,22 +61,22 @@ foreach( source ${SOURCES} )
          "${mo}"
    )
 
-   # Always make sure there's an empty English folder.  Otherwise, some menu
-   # items will be translated to the next language listed in System
-   # Preferences/Language & Text
-   if( CMAKE_SYSTEM_NAME MATCHES "Darwin" )
-      add_custom_command(
-         COMMAND
-            ${CMAKE_COMMAND} -E make_directory "${_DEST}/Resources/en.lproj"
-         OUTPUT
-            "${mo}"
-         APPEND
-      )
-   endif()
-
    list( APPEND OUTPUTS ${mo} )
 endforeach()
 
+# Always make sure there's an empty English folder.  Otherwise, some menu
+# items will be translated to the next language listed in System
+# Preferences/Language & Text
+if( CMAKE_SYSTEM_NAME MATCHES "Darwin" )
+   add_custom_command(
+      COMMAND
+         ${CMAKE_COMMAND} -E make_directory "${_DEST}/Resources/en.lproj"
+      OUTPUT
+         "${_DEST}/Resources/en.lproj/audacity.mo"
+      APPEND
+   )
+endif()
+
 add_custom_target( "${TARGET}" ALL DEPENDS ${OUTPUTS} SOURCES "${SOURCES}" )
 
 if( NOT CMAKE_SYSTEM_NAME MATCHES "Darwin" )
```

DimitarSamarov07 commented:
Let me just clarify, you modified this part of the code and moved if after the endforeach(), is that correct?

imachug commented:
> Let me just clarify, you modified this part of the code and moved if after the endforeach(), is that correct?

I moved that part out of foreach and also removed `APPEND`. I'm not exactly sure if `APPEND` should be there or not, or why the code worked at all, for that matter. Let's try to remove `APPEND` and if that doesn't work try to add `APPEND` back.

DimitarSamarov07 commented:
> > Let me just clarify, you modified this part of the code and moved if after the endforeach(), is that correct?
> 
> I moved that part out of foreach and also removed `APPEND`. I'm not exactly sure if `APPEND` should be there or not, or why the code worked at all, for that matter. Let's try to remove `APPEND` and if that doesn't work try to add `APPEND` back.

Alright, I'm gonna apply the changes in a minute, push it so I don't have to deal with Github Actions locally(+ i can always rebase out the commit if needed).

DimitarSamarov07 commented:
@imachug And now we wait.

imachug commented:
I think you forgot to remove `APPEND` so let's wait for CI and if it doesn't work try to remove it.

DimitarSamarov07 commented:
> I think you forgot to remove `APPEND` so let's wait for CI and if it doesn't work try to remove it.

Left it there on purpose to see if it works without removing it. I'm trying to not modify the build file by too much.

imachug commented:
It doesn't, let's remove it.

DimitarSamarov07 commented:
Removed it. Here we go again.

DimitarSamarov07 commented:
> It doesn't, let's remove it.

Alright, this time it didn't fail on config. Better wait to the end but I am pretty sure that fixed it

imachug commented:
It seems like it worked. If you like nice history you could squash the two top commits together but other than that LGTM.

DimitarSamarov07 commented:
@imachug Sorry, rebasing and squashing some of the commits dismissed your review. Could you please re-do it?

emabrey commented:
I need you to run `git pull` and merge the upstream changes. I made a ton of changes to fix the removal of a bunch of locales, and that is making a ton of merge conflicts.

![image](https://user-images.githubusercontent.com/3370875/125569570-c1eb5795-649c-441d-a930-740ee4cbcb95.png)

imachug commented on locale/LINGUAS:
```diff
@@ -43,8 +43,8 @@ pt_PT
 ru
 sk
-sl
 sr
+sl
```
These two got swapped accidentally I think?

DimitarSamarov07 replied:
I think so. No that it'll make a difference. I can put them in the correct order if you insist.

imachug replied:
I think they are (mostly) ordered alphabetically now, so it would be nice if it was sorted. Thanks for your work!

imachug replied:
Hi, would you please fix this or do you want me to do that?

DimitarSamarov07 replied:
Currently on a bus to home, writing from the Github mobile app. I should be home in 10-20 minutes at max, so I will fix it as soon as I get home.

DimitarSamarov07 replied:
Alright, they just announced there's gonna be a delay because of traffic. I guess I should be home in an hour. Really sorry for the inconvenience.

DimitarSamarov07 replied:
Sorry for the delay, fixed the LINGUAS order.

imachug commented:
@tenacityteam/maintainers

Can someone who can read others' mind please take a look at this? Story: last time the build failed I checked CMakeLists.txt and found this hack:

https://github.com/tenacityteam/tenacity/blob/1eba71b378c6cccf8f4f93c89ea3ef3663b61211/locale/CMakeLists.txt#L64-L75

The problem is that this code is executed *inside a foreach loop*, so it creates the exact same directory several times with a seemingly unrelated `OUTPUT` value and `APPEND` to stop CMake from arguing about invalid config.

I proposed moving this out of foreach (as you can see in this PR) to make it look at least a bit less hacky, but I can't believe that Audacity developers would use such a hack if there wasn't a reason for that. Can anyone figure out why any sane developer would do that?

DimitarSamarov07 commented:
> @tenacityteam/maintainers
> 
> Can someone who can read others' mind please take a look at this? Story: last time the build failed I checked CMakeLists.txt and found this hack:
> 
> https://github.com/tenacityteam/tenacity/blob/1eba71b378c6cccf8f4f93c89ea3ef3663b61211/locale/CMakeLists.txt#L64-L75
> 
> The problem is that this code is executed *inside a foreach loop*, so it creates the exact same directory several times with a seemingly unrelated `OUTPUT` value and `APPEND` to stop CMake from arguing about invalid config.
> 
> I proposed moving this out of foreach (as you can see in this PR) to make it look at least a bit less hacky, but I can't believe that Audacity developers would use such a hack if there wasn't a reason for that. Can anyone figure out why any sane developer would do that?

It's important to add that the previous version of the mentioned above file caused only the macOS_Intel build to fail, while the Windows 32, 64, Ubuntu and Arch builds went through without any errors. I think that someone should definitely take a look at this, I personally can't find a reason as to why(but I am also not that experienced with Cmake).

imachug commented:
> It's important to add that the previous version of the mentioned above file caused only the macOS_Intel build to fail, while the Windows 32, 64, Ubuntu and Arch builds went through without any errors.

It makes sense because there's an `if`.

DimitarSamarov07 commented:
> > It's important to add that the previous version of the mentioned above file caused only the macOS_Intel build to fail, while the Windows 32, 64, Ubuntu and Arch builds went through without any errors.
> 
> It makes sense because there's an `if`.

I think that the question we should be asking is **What did we change in the locale files that caused this build error to occur in the first place, or was it always there ?**

DimitarSamarov07 commented:
Conflicts were resolved. All I have left to do is revise all the files and make sure I haven't missed anything.

DimitarSamarov07 commented:
@imachug @emabrey 

https://github.com/tenacityteam/tenacity/blob/0221f74e774725b4be8074d0c53149e2ec25d095/locale/audacity.pot#L720


Would you prefer me to keep this as _DarkAudacity_ or shall I rename it to _DarkTenacity_ ?

imachug commented:
Let's keep 'DarkAudacity' for now. DarkAudacity is also somewhat a fork of Audacity, so I think we'll either merge its features in or remove some of them, and after that we'll remove any references to DarkAudacity. Either way, there will be no DarkTenacity or anything like that.

DimitarSamarov07 commented:
> Let's keep 'DarkAudacity' for now. DarkAudacity is also somewhat a fork of Audacity, so I think we'll either merge its features in or remove some of them, and after that we'll remove any references to DarkAudacity. Either way, there will be no DarkTenacity or anything like that.

Alright then. Made some final changes to the files, let's see if the build passes.

DimitarSamarov07 commented:
@imachug @emabrey 

I've done my final revision on the files. Sorry that all of this took so long, I just wanted to make sure I do a good job at replacing everything and only change what I need to and not mess with stuff like the commented file paths and stuff, so I reviewed every single one of the changes made with the find and replace function. That's the main reason it all took so long. In the end, all that effort was worth it as the files are finally complete. Checks are passing, so the builds are definitely working fine. Let me know if you want me to change anything else.

DimitarSamarov07 commented:
Hey,
Could someone please review my PR and let me know if I need to change anything? Thanks

Sincerely, 
Dimitar

DimitarSamarov07 commented:
@fossdd @emabrey @panos @imachug 

Could someone please take a look and see why are the builds failing? Maybe it's an issue in master as it happened after I made a merge after resolving the conflicts. Would be thankful if someone can take a look. 100% it isn't an issue with my changes, you can see for yourself:
```[ 13%] Building C object cmake-proxies/libsndfile/CMakeFiles/libsndfile.dir/__/__/lib-src/libsndfile/src/voc.c.o
[ 13%] Building C object cmake-proxies/libsndfile/CMakeFiles/libsndfile.dir/__/__/lib-src/libsndfile/src/vox_adpcm.c.o
[ 13%] Building C object cmake-proxies/libsndfile/CMakeFiles/libsndfile.dir/__/__/lib-src/libsndfile/src/w64.c.o
[ 13%] Building C object cmake-proxies/libsndfile/CMakeFiles/libsndfile.dir/__/__/lib-src/libsndfile/src/wav.c.o
[ 14%] Building C object cmake-proxies/libsndfile/CMakeFiles/libsndfile.dir/__/__/lib-src/libsndfile/src/wve.c.o
[ 14%] Building C object cmake-proxies/libsndfile/CMakeFiles/libsndfile.dir/__/__/lib-src/libsndfile/src/wavlike.c.o
[ 14%] Building C object cmake-proxies/libsndfile/CMakeFiles/libsndfile.dir/__/__/lib-src/libsndfile/src/xi.c.o
[ 14%] Linking C static library liblibsndfile.a
[ 14%] Built target libsndfile
make: *** [all] Error 2
Makefile:155: recipe for target 'all' failed
Error: Process completed with exit code 2.```

SFR-git commented:
The sourcehut build seems to have failed [even earlier](https://builds.sr.ht/~tenacity/job/548915#task-make-66).

DimitarSamarov07 commented:
> The sourcehut build seems to have failed [even earlier](https://builds.sr.ht/~tenacity/job/548915#task-make-66).

If it was caused by the locales, it would've failed at the later stages of the build or the most likely option, during the configuration part, but this early - I doubt the locales are the issue.

SFR-git commented:
> I doubt the locales are the issue.

The locales are the issue.

```
/home/build/tenacity/locale/de.po:4327:2: syntax error
```

DimitarSamarov07 commented:
> > I doubt the locales are the issue.
> 
> The locales are the issue.
> 
> ```
> /home/build/tenacity/locale/de.po:4327:2: syntax error
> ```

Well, that's surprising, let me take a look.

SFR-git commented:
Seems like an unfinished git merge artifact

Semisol commented:
Failure caused by a merge conflict artifact on `4327:2`
![image](https://user-images.githubusercontent.com/45574030/126370219-c364a81a-bbf9-461c-bbc4-6a926f6718c5.png)
Can you look into that and maybe remove the duplicate lines? Thanks.

Semisol commented:
(and please check other files too.)

SFR-git commented:
Could you `fixup` the commit into the one that caused this error?

Semisol commented:
For some reason when I built it it showed a German version and I didn't see anything referring to Tenacity...

DimitarSamarov07 commented:
> For some reason when I built it it showed a German version and I didn't see anything referring to Tenacity...

Hmm, I don't see why that would happen, you can see that all the Audacity references in the german locale have been changed, so it should use them? I'll try and look into that but I see no reason as to why that would  happen.

DimitarSamarov07 commented:
> For some reason when I built it it showed a German version and I didn't see anything referring to Tenacity...

Can I know what system are you testing on? Windows? Mac? Linux(and if which distro)?  I want to test it in the same environment + I get a segmentation fault when I try to run the built tenacity(no matter if it's master or this PR) on Debian 11, so yeah let me know what's your setup so I can try it in a VM. Thanks.

aeiouaeiouaeiouaeiouaeiouaeiou commented:
The "About" window contains an ampersand in title if the Russian localization is selected.
![2021-07-22_11-59-19](https://user-images.githubusercontent.com/69355938/126614720-f861ef7c-fb9d-4acc-ba30-7f6528da0de5.png)

Semisol commented:
> > For some reason when I built it it showed a German version and I didn't see anything referring to Tenacity...
> 
> Can I know what system are you testing on? Windows? Mac? Linux(and if which distro)? I want to test it in the same environment + I get a segmentation fault when I try to run the built tenacity(no matter if it's master or this PR) on Debian 11, so yeah let me know what's your setup so I can try it in a VM. Thanks.

MacOS.

SFR-git commented:
What's the status of this PR? Is it even relevant anymore?

DimitarSamarov07 commented:
> What's the status of this PR? Is it even relevant anymore?

Not sure if it's still relevant. No one seemed to care enough to review and approve it when it was surely relevant. If it's still relevant I can translate the files again, resolving all the conflicts. It's up to the maintainers to decide.

SFR-git commented:
@emabrey: Could you take a look at this? I'm not sure if it's already been resolved

emabrey commented:
I didn't merge this before because you can't just change the `msgid` values without also changing them in the code base. Nevertheless, I appreciate the effort. We have already handled this in English, and the `msgstr` translations need to be handled on Weblate for non-English locales.
